runtime/protos/runtime.proto

```diff
@@ -17,232 +17,330 @@ option go_package = "github.com/ServiceWeaver/weaver/runtime/protos";
+//                      │ ──────────────────────> │ GetListenerAddressReply
+//                      │                         │
+//                   3. │ <────────────────────── │ LogEntry
+//                      │ <────────────────────── │ TraceSpans
+//                      :                         :
+//
+// Note that these three types of communication can intersperse.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                      │ <────────────────────── │ EnsureComponentRequest
+//                      │ ──────────────────────> │ UpdateComponentsRequest
+//                      │ <────────────────────── │ UpdateComponentsReply
+//                      │ <────────────────────── │ LogEntry
+//                      │ ──────────────────────> │ UpdateRoutingInfoRequest
+//                      │ <────────────────────── │ UpdateRoutingInfoReply
+//                      │ ──────────────────────> │ EnsureComponentReply
+//                      :                         :
+//
+// To avoid deadlocks, a weavelet is guaranteed to process an RPC without
+// blocking and without touching the pipe.
+
+// EnvelopeMsg is a message sent by an envelope to a weavelet.
 message EnvelopeMsg {
-  // Unique id identifies a request that is part of a synchronous call. A
-  // request with positive id x expects a reply with negative id -x.
-  // Non-synchronous calls have an id of 0.
+  // TODO(mwhittaker): Don't make UpdateRoutingInfoRequest and
+  // UpdateComponentsRequest RPCs? They can be one-directional messages like
+  // logs and traces.
+
+  // A unique id for synchronous calls. A request with positive id x expects a
+  // reply with negative id -x. Non-synchronous calls have an id of 0.
   int64 id = 1;
 
   // Exactly one of the following should be set.
-  // (Avoiding protobuf oneofs since they are annoying to use.)
-
-  // A request to the weavelet to send its current metrics.
-  bool send_metrics = 2;
-
-  // A request to the weavelet to send its health status.
-  bool send_health_status = 3;
-
-  // A request to the weavelet to send the latest load information.
-  bool send_load_info = 11;
-
-  // A request to the weavelet to profile itself and send back the profile
-  // data.
-  RunProfiling run_profiling = 4;
-
-  // Setup information sent to the weavelet. This is the first message sent to
-  // the weavelet on the connection. Upon receiving the message, the weavelet
-  // immediately replies with WeaveletInfo.
-  WeaveletSetupInfo weavelet_setup_info = 5;
 
-  // A reply to the weavelet with the latest routing info.
-  RoutingInfo routing_info = 6;
+  // Handshake.
+  EnvelopeInfo envelope_info = 2;
 
-  // A reply to the weavelet with the latest list of components to start.
-  ComponentsToStart components_to_start = 7;
+  // Envelope initiated RPC requests.
+  GetHealthRequest get_health_request = 100;
```

Comment:
I think we should add these ids in order 1 to x. This is what I think people usually do.